Content.Shared/Weapons/Ranged/Systems/SharedGunSystem.Ballistic.cs

```diff
@@ -49,10 +49,48 @@ private void OnBallisticInteractUsing(EntityUid uid, BallisticAmmoProviderCompon
             return;
 
         if (GetBallisticShots(component) >= component.Capacity)
+        {
+            Popup(Loc.GetString("gun-ballistic-transfer-target-full", ("entity", args.Target)), uid, args.User);
             return;
+        }
+
+        if (EntityManager.HasComponent<SpeedLoaderComponent>(args.Used))
+        {
+            var emptySlots = component.Capacity - component.UnspawnedCount - component.Entities.Count; // Number of empty slots in the shotgun
+
+            var ammo = new List<(EntityUid? Entity, IShootable Shootable)>();
+            var ev = new TakeAmmoEvent(emptySlots, ammo, Transform(uid).Coordinates, args.User);
+            RaiseLocalEvent(args.Used, ev);
+
+            // empty speedloader
+            if (ev.Ammo.Count == 0)
+            {
+                Popup(Loc.GetString("gun-speedloader-empty"), uid, args.User);
+                return;
+            }
+
+            foreach (var (ent, _) in ammo)
+            {
+                if (ent == null)
+                    continue;
+
+                component.Entities.Add(ent.Value);
+                Containers.Insert(ent.Value, component.Container);
+
+                if (IsClientSide(ent.Value))
+                    Del(ent.Value);
+
```

Comment:
Why is this inserting then deleting it?

Reply:
This is what happens without the delete. `OnBallisticAmmoFillDoAfter` does the same thing

https://github.com/user-attachments/assets/1346e23b-ce2e-4141-a1ed-6a80cc6d83eb